micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java

```diff
@@ -0,0 +1,279 @@
+     *
+     * @param kafkaStreams instance to be instrumented
+     * @param tags         additional tags
+     */
+    public KafkaMetrics(KafkaStreams kafkaStreams, Iterable<Tag> tags) {
+        this(kafkaStreams::metrics, tags);
+    }
+
+    /**
+     * {@link KafkaStreams} metrics binder
+     *
+     * @param kafkaStreams instance to be instrumented
+     */
+    public KafkaMetrics(KafkaStreams kafkaStreams) {
+        this(kafkaStreams::metrics);
+    }
+
+    /**
+     * Kafka {@link AdminClient} metrics binder
+     *
+     * @param adminClient instance to be instrumented
+     * @param tags        additional tags
+     */
+    public KafkaMetrics(AdminClient adminClient, Iterable<Tag> tags) {
+        this(adminClient::metrics, tags);
+    }
+
+    /**
+     * Kafka {@link AdminClient} metrics binder
+     *
+     * @param adminClient instance to be instrumented
+     */
+    public KafkaMetrics(AdminClient adminClient) {
+        this(adminClient::metrics);
+    }
+
+    KafkaMetrics(Supplier<Map<MetricName, ? extends Metric>> metricsSupplier) {
+        this(metricsSupplier, emptyList());
+    }
+
+    KafkaMetrics(Supplier<Map<MetricName, ? extends Metric>> metricsSupplier,
+            Iterable<Tag> extraTags) {
+        this.metricsSupplier = metricsSupplier;
+        this.extraTags = extraTags;
+    }
+
+    @Override
+    public void bindTo(MeterRegistry registry) {
+        checkAndBindMetrics(registry);
+    }
+
+    /**
+     * Gather metrics from Kafka metrics API and register Meters.
+     * <p>
+     * As this is a one-off execution when binding a Kafka client, Meters include a call to this
+     * validation to double-check new metrics when returning values. This should only add the cost of
+     * validating meters registered counter when no new meters are present.
+     */
+    void checkAndBindMetrics(MeterRegistry registry) {
+        Map<MetricName, ? extends Metric> metrics = metricsSupplier.get();
+        //Only happens first time number of metrics change
+        if (currentSize.get() != metrics.size()) {
+            currentSize.set(metrics.size());
+            Map<String, Set<Meter>> boundMeters = new HashMap<>();
+            //Register meters
+            metrics.forEach((name, metric) -> {
+                //Filter out metrics from group "app-info", that includes metadata
+                if (METRIC_GROUP_APP_INFO.equals(name.group())) {
+                    currentSize.incrementAndGet();
+                    return;
+                }
+                Meter meter = bindMeter(registry, metric);
+                //Collect metrics with same name to validate number of labels
+                Set<Meter> meters = boundMeters.get(metric.metricName().name());
+                if (meters == null) meters = new HashSet<>();
+                meters.add(meter);
+                boundMeters.put(metric.metricName().name(), meters);
+            });
+
+            //Remove meters with lower number of tags
```

Comment:
Could we prevent both from being registered at the same time? That is, remove the meters with the lower number of tags before adding the meter with more tags? As it is, there is a race condition where meters could be retrieved (published/polled/scraped) with more than one meter registered with the same name but different tags.

Reply:
You're right. It could be safer to remove all registered meters before starting a new round, and synchronize this block in order to avoid race conditions. Assuming there are no meters with different number of tags on the same call--my experience is that this diff happens _between_ calls, removing upfront meters will avoid the need for future filtering. I've added a refactor to experiment with this.